src/plugins/mvcc_delete_plugin.cpp

```diff
@@ -0,0 +1,177 @@
+#include "mvcc_delete_plugin.hpp"
+
+#include "operators/get_table.hpp"
+#include "operators/table_wrapper.hpp"
+#include "operators/update.hpp"
+#include "operators/validate.hpp"
+#include "storage/pos_list.hpp"
+#include "storage/reference_segment.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+MvccDeletePlugin::MvccDeletePlugin()
+    : _sm(StorageManager::get()),
+      _delete_threshold_rate_invalidated_rows(0.8),
+      _delete_threshold_commit_diff_factor(1.5),
+      _idle_delay_logical_delete(std::chrono::milliseconds(1000)),
+      _idle_delay_physical_delete(std::chrono::milliseconds(1000)) {}
+
+const std::string MvccDeletePlugin::description() const { return "Physical MVCC delete plugin"; }
+
+void MvccDeletePlugin::start() {
+  _loop_thread_logical_delete = std::make_unique<PausableLoopThread>(_idle_delay_logical_delete,
+                                                                     [this](size_t) { this->_logical_delete_loop(); });
+
+  _loop_thread_physical_delete = std::make_unique<PausableLoopThread>(
+      _idle_delay_physical_delete, [this](size_t) { this->_physical_delete_loop(); });
+}
+
+void MvccDeletePlugin::stop() {
+  // Call destructor of PausableLoopThread to terminate its thread
+  _loop_thread_logical_delete.reset();
+  _loop_thread_physical_delete.reset();
+}
+
+/**
+ * This function analyzes each chunk of every table and triggers a chunk-cleanup-procedure if a certain threshold of invalidated rows is exceeded.
+ */
+void MvccDeletePlugin::_logical_delete_loop() {
+  for (auto& [table_name, table] : _sm.tables()) {
+    // Check all chunks, except for the last one, which is currently used for insertions
+    if (table->chunk_count() == 1) return;
+    ChunkID max_chunk_id_to_check = static_cast<ChunkID>(table->chunk_count() - 2);
+
+    for (ChunkID chunk_id = ChunkID{0}; chunk_id <= max_chunk_id_to_check; chunk_id++) {
+      const auto& chunk = table->get_chunk(chunk_id);
+      if (chunk && chunk->get_cleanup_commit_id() == MvccData::MAX_COMMIT_ID) {
+        // Calculate metric 1 – Chunk invalidation level
+        const double rate_of_invalidated_rows = static_cast<double>(chunk->invalid_row_count()) / chunk->size();
+
+        // Calculate metric 2 – Chunk Hotness
+        const CommitID lowest_end_commit_id =
+            *std::min_element(std::begin(chunk->mvcc_data()->end_cids), std::end(chunk->mvcc_data()->end_cids));
```

Comment:
This is a bit late for me to realize, but if we are looking at the CIDs anyway, what do we save by maintaining the `invalid_row_count`?

Reply:
Legit point! We should keep the `invalid_row_count` metric though, because It is very efficient. The second metric in comparison is very resource-intensive, since we're evaluating all rows over and over again. This should only be done, if the first criterion (threshold `invalid_row_count` surpassed) becomes positive.

Reply:
Yes, keeping it, but only evaluating it if necessary sounds good.